### PR TITLE
Refactoring preview features analyzer

### LINKED ISSUE #5850: Questionable behavior of DetectPreviewFeatures analyzers
https://github.com/dotnet/roslyn-analyzers/blob/60eb13c44c366205e5a4dec195404477009c41de/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureTests.Generics.cs#L13-L45

Diagnostic span for this test is very inconsistent with it everywhere else.

@jeffhandley Can you confirm if this behavior is correct?

### PR DESCRIPTION
**Pull request by Youssef1313**
I'm feeling the approach I'm taking here is more clearer and less error-prone. @jeffhandley Let me know if you want me to continue the refactoring this way.

Closes #5850

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs [hunk 1]
```diff
@@ -1081,5 +1054,44 @@ private static bool SymbolIsAnnotatedAsPreview(ISymbol symbol, ConcurrentDiction
 
             return existing.isPreview;
         }
+
+        protected void AnalyzeTypeSyntax(SyntaxNodeAnalysisContext context, SyntaxNode node, ConcurrentDictionary<ISymbol, (bool isPreview, string? message, string? url)> requiresPreviewFeaturesSymbols, Func<ISymbol, bool> symbolIsAnnotatedAsPreview)
+        {
+            var type = context.SemanticModel.GetTypeInfo(node, context.CancellationToken).Type;
+            if (type is not null && symbolIsAnnotatedAsPreview(type))
+            {
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:1062]: This is the core of the refactoring. I'm moving reporting diagnostics to this method.

Here, I'm analyzing `NameSyntax`, then referring back to `SemanticModel` to get the symbol I'm dealing with, and determine if we have a special diagnostic for it.

**Comment by Youssef1313** [comment]
@jeffhandley Can you take a look at the updated tests and validate if the new behavior for a couple of tests is desired?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs [hunk 2]
```diff
@@ -20,7 +20,17 @@ namespace Microsoft.NetCore.Analyzers.Runtime
...
+        TBaseTypeOrImplementsOrInherits,
+        TBaseTypeDeclarationSyntax,
+        TTypeConstraintSyntax,
+        TTypeArgumentList,
+        TParameterSyntax> : DiagnosticAnalyzer
+        where TBaseTypeOrImplementsOrInherits : SyntaxNode
+        where TBaseTypeDeclarationSyntax : SyntaxNode
+        where TTypeConstraintSyntax : SyntaxNode
+        where TTypeArgumentList : SyntaxNode
+        where TParameterSyntax : SyntaxNode
```
Comment (buyaa-n) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs:33]: I might miss something, seems this and `TBaseTypeDeclarationSyntax`  are not used

**Review by buyaa-n — COMMENTED** [review]
> Can you take a look at the updated tests and validate if the new behavior for a couple of tests is desired

Sorry for late reply, the test updates so far looks good to me, i realize the code refactoring is not complete yet, so did not review completely, please continue your refactoring, thanks!

**Comment by Youssef1313** [comment]
> i realize the code refactoring is not complete yet

Yes. Unfortunately the last time I tried to complete this it was tricky to get everything working as intended :(

I'll give it another look and see what I can get.

**Comment by jeffhandley** [comment]
I'm closing this draft PR; we can use a fresh PR if this is revisited.

**Pull request closed without merging**